File: backend/src/ingestion/filters.ts
import { SkipIngest } from "./errors.js";

export { SkipIngest };

/** Shown on the Jobs page when a site serves a bot-protection challenge instead of content. */
export const BOT_BLOCK_SKIP_MESSAGE =
  "Site blocked automated access (Cloudflare or bot protection page)";

const BOT_BLOCK_HTML_PATTERNS = [
  /just a moment/i,
  /enable javascript and cookies to continue/i,
  /please enable js(?:\s+and disable any ad blocker)?/i,
  /challenges\.cloudflare\.com/i,
  /cf-browser-verification/i,
  /checking your browser/i,
  /attention required!?\s*\|\s*cloudflare/i,
  /cf-challenge/i,
  /ddos protection by cloudflare/i,
  /performing security verification/i,
  /please wait while we verify/i,
  /__cf_chl/i,
  /datadome/i,
];

const BOT_BLOCK_EXTRACTED_PATTERNS = [
  /^enable javascript and cookies to continue$/i,
  /^checking your browser before accessing/i,
  /^please enable cookies/i,
  /please enable js(?:\s+and disable any ad blocker)?/i,
];

const BOT_BLOCK_SIGNAL_RE =
  /cloudflare|javascript|cookies|challenge|verify|enable js|ad blocker|datadome/i;

/**
 * Detect Cloudflare and similar bot-protection challenge pages.
 * Returns a user-facing skip reason, or null when the page looks normal.
 */
export function detectBotBlockPage(
  html: string,
  options?: { extractedText?: string | null; httpStatus?: number },
): string | null {
  const sample = (html || "").slice(0, 12_000);
  if (sample && BOT_BLOCK_HTML_PATTERNS.some((p) => p.test(sample))) {
    return BOT_BLOCK_SKIP_MESSAGE;
  }

  const extracted = (options?.extractedText ?? "").trim();
  if (
    extracted &&
    BOT_BLOCK_EXTRACTED_PATTERNS.some((p) => p.test(extracted))
  ) {
    return BOT_BLOCK_SKIP_MESSAGE;
  }

  const status = options?.httpStatus;
  if (
    status != null &&
    [401, 403, 429].includes(status) &&
    extracted.length > 0 &&
    extracted.length < 500 &&
    BOT_BLOCK_SIGNAL_RE.test(`${sample} ${extracted}`)
  ) {
    return BOT_BLOCK_SKIP_MESSAGE;
  }

  return null;
}

// ---------- Soft 404 / empty-content detection ----------

const SOFT_404_PAT =
  /(404|not\s+found|page\s+not\s+found|page\s+removed|gone|error\s+404)/i;

/** Detect real and "soft" 404s or nav-only pages. Port of `looks_like_soft_404`. */
export function looksLikeSoft404(
  html: string,
  status: number,
  options?: { extractedText?: string | null; minTextBytes?: number },
): boolean {
  const minTextBytes = options?.minTextBytes ?? 500;
  const extractedText = options?.extractedText;

  if (status === 404 || status === 410 || status === 451) {
    return true;
  }

  if (html && SOFT_404_PAT.test(html.slice(0, 2000) || "")) {
    return true;
  }

  if (extractedText != null && extractedText.trim().length < minTextBytes) {
    return true;
  }

  return false;
}

// ---------- Fast AI-topic classifier (regex keywords; no LLM) ----------

const AI_INCLUDE = [
  /\b(openai|chatgpt|gpt-4|gpt-3|gemini|anthropic|claude|copilot|deepseek)\b/i,
  /\b(ai|artificial intelligence)\b/i,
  /\b(llms?|large language model|language model)\b/i,
  /\b(machine learning|ml|deep learning|neural network)\b/i,
  /\b(embedding|token|inference|fine[- ]?tuning|rlhf|safety|alignment)\b/i,
  /\b(prompt|prompt injection|jailbreak|hallucination|red teaming)\b/i,
  /\b(model weights?|training data|dataset|inference api)\b/i,
  /\b(privacy|pii|data leakage|governance|compliance) in (ai|ml)\b/i,
];

const AI_EXCLUDE = [
  /\b(sports|football|basketball|baseball|soccer|golf|tennis)\b/i,
  /\b(recipes?|cooking|travel|tourism|celebrity|fashion|beauty)\b/i,
  /\b(stock photos?|coupon|promo|giveaway)\b/i,
];

function hitCount(text: string, patterns: RegExp[]): number {
  const low = (text || "").toLowerCase();
  return patterns.reduce((n, p) => (p.test(low) ? n + 1 : n), 0);
}

export type ClassifyDetails = {
  include_hits: number;
  exclude_hits: number;
  threshold: number;
};

/** Job skip reason when URL/title/content matches a non-AI topic pattern. */
export const NOT_AI_RELATED_SKIP_MESSAGE = "not ai related";

function topicBlob(parts: {
  url?: string;
  title?: string;
  text?: string;
}): string {
  return [parts.title, parts.url, parts.text]
    .map((part) => part?.trim())
    .filter((part): part is string => Boolean(part))
    .join(" ")
    .trim();
}

/** Return skip reason when sports/travel/recipes/etc. appear without enough AI context. */
export function getExcludedNonAiTopicSkipReason(parts: {
  url?: string;
  title?: string;
  text?: string;
}): string | null {
  const blob = topicBlob(parts);
  if (!blob) return null;

  const exclude_hits = hitCount(blob, AI_EXCLUDE);
  if (exclude_hits === 0) return null;

  const include_hits = hitCount(blob, AI_INCLUDE);
  const threshold = 2;
  // News pages often mention Travel/Sports/Food in nav/footer chrome. Do not skip
  // when the article itself is clearly AI-related.
  if (include_hits - exclude_hits >= threshold) return null;

  const headBlob = topicBlob({ url: parts.url, title: parts.title });
  if (headBlob && hitCount(headBlob, AI_INCLUDE) >= 1) return null;

  return NOT_AI_RELATED_SKIP_MESSAGE;
}

/** Simple deterministic filter. Port of `classify_ai_related`. */
export function classifyAiRelated(
  text: string,
  options?: { title?: string; url?: string; includeThreshold?: number },
): [boolean, ClassifyDetails] {
  const title = options?.title ?? "";
  const url = options?.url ?? "";
  const threshold = options?.includeThreshold ?? 2;
  const blob = topicBlob({ title, url, text });

  const include_hits = hitCount(blob, AI_INCLUDE);
  const exclude_hits = hitCount(blob, AI_EXCLUDE);
  const ok =
    exclude_hits === 0 && include_hits - exclude_hits >= threshold;

  return [ok, { include_hits, exclude_hits, threshold }];
}

File: python/app/ingestion/filters.py
"""Port of app.ingestion.filters — soft 404 and AI-topic classification."""

from __future__ import annotations

import re
from typing import Any

from .errors import SkipIngest

BOT_BLOCK_SKIP_MESSAGE = (
    "Site blocked automated access (Cloudflare or bot protection page)"
)

BOT_BLOCK_HTML_PATTERNS = [
    re.compile(r"just a moment", re.I),
    re.compile(r"enable javascript and cookies to continue", re.I),
    re.compile(r"please enable js(?:\s+and disable any ad blocker)?", re.I),
    re.compile(r"challenges\.cloudflare\.com", re.I),
    re.compile(r"cf-browser-verification", re.I),
    re.compile(r"checking your browser", re.I),
    re.compile(r"attention required!?\s*\|\s*cloudflare", re.I),
    re.compile(r"cf-challenge", re.I),
    re.compile(r"ddos protection by cloudflare", re.I),
    re.compile(r"performing security verification", re.I),
    re.compile(r"please wait while we verify", re.I),
    re.compile(r"__cf_chl", re.I),
    re.compile(r"datadome", re.I),
]

BOT_BLOCK_EXTRACTED_PATTERNS = [
    re.compile(r"^enable javascript and cookies to continue$", re.I),
    re.compile(r"^checking your browser before accessing", re.I),
    re.compile(r"^please enable cookies", re.I),
    re.compile(r"please enable js(?:\s+and disable any ad blocker)?", re.I),
]

BOT_BLOCK_SIGNAL_RE = re.compile(
    r"cloudflare|javascript|cookies|challenge|verify|enable js|ad blocker|datadome",
    re.I,
)

NOT_AI_RELATED_SKIP_MESSAGE = "not ai related"

__all__ = [
    "SkipIngest",
    "BOT_BLOCK_SKIP_MESSAGE",
    "NOT_AI_RELATED_SKIP_MESSAGE",
    "detect_bot_block_page",
    "excluded_non_ai_topic_skip_reason",
    "looks_like_soft_404",
    "classify_ai_related",
]

SOFT_404_PAT = re.compile(
    r"(404|not\s+found|page\s+not\s+found|page\s+removed|gone|error\s+404)",
    re.I,
)

AI_INCLUDE = [
    re.compile(r"\b(openai|chatgpt|gpt-4|gpt-3|gemini|anthropic|claude|copilot|deepseek)\b", re.I),
    re.compile(r"\b(ai|artificial intelligence)\b", re.I),
    re.compile(r"\b(llms?|large language model|language model)\b", re.I),
    re.compile(r"\b(machine learning|ml|deep learning|neural network)\b", re.I),
    re.compile(
        r"\b(embedding|token|inference|fine[- ]?tuning|rlhf|safety|alignment)\b",
        re.I,
    ),
    re.compile(r"\b(prompt|prompt injection|jailbreak|hallucination|red teaming)\b", re.I),
    re.compile(r"\b(model weights?|training data|dataset|inference api)\b", re.I),
    re.compile(r"\b(privacy|pii|data leakage|governance|compliance) in (ai|ml)\b", re.I),
]

AI_EXCLUDE = [
    re.compile(r"\b(sports|football|basketball|baseball|soccer|golf|tennis)\b", re.I),
    re.compile(r"\b(recipes?|cooking|travel|tourism|celebrity|fashion|beauty)\b", re.I),
    re.compile(r"\b(stock photos?|coupon|promo|giveaway)\b", re.I),
]


def detect_bot_block_page(
    html: str,
    *,
    extracted_text: str | None = None,
    http_status: int | None = None,
) -> str | None:
    """Return a skip message when HTML looks like a bot-protection challenge page."""
    sample = (html or "")[:12_000]
    if sample and any(p.search(sample) for p in BOT_BLOCK_HTML_PATTERNS):
        return BOT_BLOCK_SKIP_MESSAGE

    extracted = (extracted_text or "").strip()
    if extracted and any(p.search(extracted) for p in BOT_BLOCK_EXTRACTED_PATTERNS):
        return BOT_BLOCK_SKIP_MESSAGE

    if (
        http_status in (401, 403, 429)
        and extracted
        and len(extracted) < 500
        and BOT_BLOCK_SIGNAL_RE.search(f"{sample} {extracted}")
    ):
        return BOT_BLOCK_SKIP_MESSAGE

    return None


def looks_like_soft_404(
    html: str,
    status: int,
    *,
    extracted_text: str | None = None,
    min_text_bytes: int = 500,
) -> bool:
    if status in (404, 410, 451):
        return True
    if html and SOFT_404_PAT.search((html or "")[:2000]):
        return True
    if extracted_text is not None and len(extracted_text.strip()) < min_text_bytes:
        return True
    return False


def _hit_count(text: str, patterns: list[re.Pattern[str]]) -> int:
    low = (text or "").lower()
    return sum(1 for p in patterns if p.search(low))


def _topic_blob(*, url: str = "", title: str = "", text: str = "") -> str:
    return " ".join(
        part.strip() for part in (title, url, text) if part and part.strip()
    ).strip()


def excluded_non_ai_topic_skip_reason(
    *,
    url: str = "",
    title: str = "",
    text: str = "",
) -> str | None:
    """Return skip reason when sports/travel/recipes/etc. appear without enough AI context."""
    blob = _topic_blob(url=url, title=title, text=text)
    if not blob:
        return None
    exclude_hits = _hit_count(blob, AI_EXCLUDE)
    if exclude_hits == 0:
        return None

    include_hits = _hit_count(blob, AI_INCLUDE)
    threshold = 2
    if include_hits - exclude_hits >= threshold:
        return None

    head_blob = _topic_blob(url=url, title=title)
    if head_blob and _hit_count(head_blob, AI_INCLUDE) >= 1:
        return None

    return NOT_AI_RELATED_SKIP_MESSAGE


def classify_ai_related(
    text: str,
    *,
    title: str = "",
    url: str = "",
    include_threshold: int = 2,
) -> tuple[bool, dict[str, Any]]:
    blob = _topic_blob(url=url, title=title, text=text)
    include_hits = _hit_count(blob, AI_INCLUDE)
    exclude_hits = _hit_count(blob, AI_EXCLUDE)
    ok = exclude_hits == 0 and (include_hits - exclude_hits) >= include_threshold
    details = {
        "include_hits": include_hits,
        "exclude_hits": exclude_hits,
        "threshold": include_threshold,
    }
    return ok, details
